core/src/main/java/org/apache/iceberg/avro/ProjectionDatumReader.java

```diff
@@ -59,7 +71,9 @@ public void setRowPositionSupplier(Supplier<Long> posSupplier) {
 
   @Override
   public void setSchema(Schema newFileSchema) {
-    this.fileSchema = newFileSchema;
+    if (this.fileSchema == null) {
+      this.fileSchema = newFileSchema;
+    }
```

Comment:
It seems we are ignoring `newFileSchema` in the uniontype case, which might be an incorrect use of the API. Note that this method overrides its parent in Avro's `DatumReader`.